hypothesis-python/src/hypothesis/internal/conjecture/engine.py

```diff
@@ -378,16 +378,26 @@ def _cache(self, data: ConjectureData) -> None:
             self.__data_cache_ir[key] = result
 
     def cached_test_function_ir(
-        self, nodes: List[IRNode]
+        self, nodes: List[IRNode], *, error_on_discard: bool = False
     ) -> Union[ConjectureResult, _Overrun]:
         key = self._cache_key_ir(nodes=nodes)
         try:
             return self.__data_cache_ir[key]
         except KeyError:
             pass
 
+        observer: Optional[DataObserver] = None
```

Comment:
this is a subtle semantics change from `cached_test_function`, which set `observer=DataObserver()` (a no-op) in the `error_on_discard=False` case. I don't see a reason not to use the datatree observer to add observations back to the tree.

Maybe the idea is `simulate_test_function` can't possibly explore new paths, so using a no-op observer is a performance win?

Reply:
Agree it's probably for speed.

Reply:
makes sense! I've changed to DataObserver here.